sklearn/calibration.py

```diff
@@ -311,145 +370,202 @@ def _more_tags(self):
-            self.label_encoder_.fit(y)
+def _fit_calibrator(clf_fitted, preds, y, label_encoder_, method,
+                    sample_weight=None):
+    """Fit calibrator(s) and return a `_CalibratedClassiferPipeline`
+    instance.
+
+    `n_classes` (i.e. `len(clf_fitted.classes_)`) calibrators are fitted.
+    However, if `n_classes` equals 2, one calibrator is fit.
+
+    Parameters
+    ----------
+    clf_fitted : Estimator instance
+        Fitted classifier.
+
+    preds :  array-like, shape (n_samples, n_classes)
+        Predictions for calibrating the predictions.
+        If binary, shape (n_samples, 1).
+
+    y : ndarray, shape (n_samples,)
+        The targets.
+
+    label_encoder_ : LabelEncoder instance
+        LabelEncoder instance fitted on all the targets.
+
+    method : {'sigmoid', 'isotonic'}
+        The method to use for calibration.
+
+    sample_weight : ndarray, shape (n_samples,), default=None
+        Sample weights. If None, then samples are equally weighted.
+
+    Returns
+    -------
+    pipeline : _CalibratedClassiferPipeline instance
+    """
+    Y = label_binarize(y, classes=label_encoder_.classes_)
+    pos_class_indices = label_encoder_.transform(clf_fitted.classes_)
+    calibrated_classifiers = []
+    for class_idx, this_pred in zip(pos_class_indices, preds.T):
+        if method == 'isotonic':
+            calibrator = IsotonicRegression(out_of_bounds='clip')
+        elif method == 'sigmoid':
+            calibrator = _SigmoidCalibration()
         else:
-            self.label_encoder_.fit(self.classes)
+            raise ValueError("'method' should be one of: 'sigmoid' or "
+                             f"'isotonic'. Got {method}.")
+        calibrator.fit(this_pred, Y[:, class_idx], sample_weight)
+        calibrated_classifiers.append(calibrator)
 
-        self.classes_ = self.label_encoder_.classes_
-        Y = label_binarize(y, classes=self.classes_)
+    pipeline = _CalibratedClassiferPipeline(
+        clf_fitted, calibrated_classifiers, label_encoder_
+    )
+    return pipeline
 
-        df, idx_pos_class = self._preproc(X)
-        self.calibrators_ = []
 
-        for k, this_df in zip(idx_pos_class, df.T):
-            if self.method == 'isotonic':
-                calibrator = IsotonicRegression(out_of_bounds='clip')
-            elif self.method == 'sigmoid':
-                calibrator = _SigmoidCalibration()
-            else:
-                raise ValueError('method should be "sigmoid" or '
-                                 '"isotonic". Got %s.' % self.method)
-            calibrator.fit(this_df, Y[:, k], sample_weight)
-            self.calibrators_.append(calibrator)
+class _CalibratedClassiferPipeline:
+    """Pipeline-like chaining a fitted classifier and its fitted calibrators.
 
-        return self
+    Parameters
+    ----------
+    clf_fitted : Estimator instance
+        Fitted classifier.
+
+    calibrators_fitted : List of fitted estimator instances
```

Comment:
nit: I'd be fine calling these `clf` and `calibrators` since the doc is clear, but as you prefer.